File: nreduce/benchmark/mergesort.pl
#!/usr/bin/perl -w

sub genlist
{
  my ($max) = @_;
  my @res = ();

  for (my $i = 0; $i < $max; $i++) {
    my $str = 7*$i;
    my $rev = reverse($str);
    push(@res,$rev);
  }

  return @res;
}

sub mergesort
{
  my @m = @_;

  if (scalar(@m) <= 1) {
    return @m;
  }

  my $middle = scalar(@m)/2;
  my @left = @m[0..$middle-1];
  my @right = @m[$middle..$#m];

  @left = mergesort(@left);
  @right = mergesort(@right);

  return merge(\@left,\@right);
}

sub merge {
  my ($xref,$yref) = @_;
  my @xlst = @{$xref};
  my @ylst = @{$yref};
  my $x = 0;
  my $y = 0;
  my @res = ();

  while (($x < scalar(@xlst)) && ($y < scalar(@ylst))) {
    my $xval = $xlst[$x];
    my $yval = $ylst[$y];
    if ($xval lt $yval) {
      push(@res,$xval);
      $x++;
    }
    else {
      push(@res,$yval);
      $y++;
    }
  }
  if ($x < scalar(@xlst)) {
    push(@res,@xlst[$x..$#xlst]);
  }
  if ($y < scalar(@ylst)) {
    push(@res,@ylst[$y..$#ylst]);
  }
  return @res;
}

my $n = ($#ARGV < 0) ? 1000 : $ARGV[0];
my @input = genlist($n);
my @sorted = mergesort(@input);
foreach $str (@sorted) {
  print "$str\n";
}

File: nreduce/benchmark/mergesort.js
function genlist(max)
{
  var res = new Array(max);
  for (var i = 0; i < max; i++) {
    var chars = (7*i)+"";
    var rev = "";
    for (var c = 0; c < chars.length; c++)
      rev += chars[chars.length-1-c];
    res[i] = rev;
  }
  return res;
}

function mergesort(m)
{
  if (m.length <= 1)
    return m;

  var middle = m.length/2;
  var left = m.slice(0,middle);
  var right = m.slice(middle,m.length);

  left = mergesort(left);
  right = mergesort(right);

  return merge(left,right);
}

function merge(xlst,ylst)
{
  var x = 0;
  var y = 0;
  var res = new Array();
  var rpos = 0;
  while ((x < xlst.length) && (y < ylst.length)) {
    var xval = xlst[x];
    var yval = ylst[y];
    if (xval < yval) {
      res.push(xval);
      x++;
    }
    else {
      res.push(yval);
      y++;
    }
  }
  if (x < xlst.length)
    res = res.concat(xlst.slice(x,xlst.length));
  if (y < ylst.length)
    res = res.concat(ylst.slice(y,ylst.length));
  return res;
}

var n = 1000;
if (arguments.length > 0)
  n = arguments[0];
var strings = genlist(n);
var sorted = mergesort(strings);
for (var i = 0; i < sorted.length; i++)
  print(sorted[i]);
